Fix: Clean up FAB permissions when deleting DAGs

HsiuChuanHsu opened the pull request:
<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
## Description
When a DAG is deleted, the corresponding Flask-AppBuilder (FAB) permissions remain in the database tables (`ab_view_menu`, `ab_permission_view`, `ab_permission_view_role`). This causes
1. **Orphaned permissions**: Deleted DAGs continue to appear in the permission management UI
2. **User confusion**: Users see permissions for non-existent DAGs

## How this happend?
The `delete_dag()` function in `delete_dag.py` successfully removes DAG-related records from core Airflow tables but does not clean up the Flask-AppBuilder permission system.

## Solution
- **FAB Provider**
  - `dag_permissions.py`: New module that handles FAB permission cleanup ( code referenced from https://github.com/apache/airflow/blob/e00777aa13556b3a81bec00dab50bcfbf98a3d86/devel-common/src/tests_common/test_utils/db.py#L358-L393)
  - `fab_auth_manager.py`: Added `cleanup_dag_permissions()` method
  - `Tests`: New tests for the permission cleanup logic

- **Core Airflow (delete_dag.py)**
  - Now calls `auth_manager.cleanup_dag_permissions()`
  - Added error handling so DAG deletion doesn't fail if permission cleanup fails


## Replace automatic DAG permission cleanup with CLI command
- **Changes:**
  - Remove automatic permission cleanup from `delete_dag()` function
  - Add new `airflow fab-auth-manager permissions-cleanup` CLI command
  - Support both global orphaned permission cleanup and specific DAG cleanup 

- **CLI command supports**  
  - Cleaning up orphaned permissions for deleted DAGs (default behavior)
  - Cleaning up permissions for specific DAG with `--dag-id` option
  - Dry-run mode with `--dry-run` flag
  - Skip confirmation with `--yes` flag
  - Verbose output with `--verbose` flag

- **Examples**:
  - `airflow fab-auth-manager permissions-cleanup`
  - `airflow fab-auth-manager permissions-cleanup --dry-run`
  - `airflow fab-auth-manager permissions-cleanup --dag-id my_dag`
  - `airflow fab-auth-manager permissions-cleanup --yes`


closes: #50905 

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [airflow-core/newsfragments](https://github.com/apache/airflow/tree/main/airflow-core/newsfragments).

HsiuChuanHsu committed:
Fix: Clean up FAB permissions when deleting DAGs (#50905)
This commit adds cleanup logic to the delete_dag function to:
- Remove DAG-specific resources from ab_view_menu table
- Clean up associated permissions and role associations

HsiuChuanHsu committed:
test: Add comprehensive tests for DAG permission cleanup
Test Coverage:
- Non-FAB auth manager scenarios (graceful skip)
- FAB provider unavailable scenarios (import error handling)
- Full cleanup process with mocked FAB models
- Integration with delete_dag function
- Edge cases (non-existent DAGs)

pierrejeambrun commented on airflow-core/src/airflow/api/common/delete_dag.py:
```diff
@@ -88,4 +88,107 @@ def delete_dag(dag_id: str, keep_records_in_log: bool = True, session: Session =
+
+                return fab_models
+            except ImportError:
+                # If FAB provider is not available, skip cleanup
```
Maybe this should be offloaded to the providers by using a generic 'cleanup' interface callable from core. This way we do not have provider specific code in here and each auth manager is free to implement its own version of `clear_permissions(...)`

HsiuChuanHsu replied:
Thanks for the feedback! That makes a lot of sense.

Offloading this to the providers would make the core code a lot cleaner and easier to maintain. I'll go ahead and implement it this way:
1.  I'll create a new module, `dag_permissions.py`, to specifically handle the DAG permission cleanup logic within the FAB provider.
2.  I'll add a `cleanup_dag_permissions()` method to the `FabAuthManager` class that delegates the task to the new module.
https://github.com/apache/airflow/blob/e00777aa13556b3a81bec00dab50bcfbf98a3d86/providers/fab/src/airflow/providers/fab/auth_manager/fab_auth_manager.py#L181
3. I'll update `delete_dag.py` to  calls the `cleanup_dag_permissions()` method on the current authentication manager.

This feels like a much better approach. I'll get started on this.

pierrejeambrun reviewed (commented):
Looking good, just a suggestion because I don't think this code should live in core.

cc: @vincbeck

HsiuChuanHsu committed:
fix: clean up DAG permissions when deleting DAGs
- : Simplified core logic using auth manager delegation

- : Added cleanup_dag_permissions method

- : New specialized module for FAB permission cleanup

- Enhanced test coverage for multiple auth manager scenarios

vincbeck commented on airflow-core/tests/unit/api/common/test_delete_dag.py:
```diff
@@ -0,0 +1,130 @@
+        try:
+            _cleanup_dag_permissions("test_dag", session)
+        except Exception:
+            pytest.fail("_cleanup_dag_permissions should handle get_auth_manager exceptions gracefully")
```
```suggestion
         _cleanup_dag_permissions("test_dag", session)
```

vincbeck reviewed (commented):
I am not against this change but here are some thoughts:
- Do we really want that? Example: As an admin, I give read permissions for the Dag `secret` to X users. This Dag gets deleted and re-created, all permissions will be lost and I (the admin) needs to assign again permissions to my users.
- Since this is very auth manager specific, and actually very Fab Auth manager specific. Should we not, instead, provide a CLI like `airflow fab-auth-manager permissions-cleanup` which would delete all orphaned permissions. That way we do not need to wire the mechanism in Airflow with something like `if hasattr(...)`, and it would be up to the admin to run this command when they think it is needed

uranusjr commented on airflow-core/src/airflow/api/common/delete_dag.py:
```diff
@@ -88,4 +88,34 @@ def delete_dag(dag_id: str, keep_records_in_log: bool = True, session: Session =
+        log.warning(
+            "Failed to clean up DAG permissions for dag_id %s: %s. DAG deletion will continue.", dag_id, e
+        )
+        pass
```
```suggestion
```

uranusjr commented on airflow-core/src/airflow/api/common/delete_dag.py:
```diff
@@ -88,4 +88,34 @@ def delete_dag(dag_id: str, keep_records_in_log: bool = True, session: Session =
+    try:
+        from airflow.api_fastapi.app import get_auth_manager
+
+        auth_manager = get_auth_manager()
```
Should these lines be expected to fail? I feel they should be moved outside the try block.

HsiuChuanHsu committed:
feat(fab-auth-manager): replace automatic DAG permission cleanup with CLI command
Replace automatic cleanup of DAG permissions during DAG deletion with a new CLI command
approach as requested in code review. This gives operators explicit control over when
and how DAG permissions are cleaned up.

HsiuChuanHsu committed:
feat(fab): Update DAG permissions cleanup to SQLAlchemy 2.0 and improve tests
- Replace deprecated session.query() with session.scalars(select()) and session.execute(delete())
- Update permissions_command.py to use SQLAlchemy 2.0 syntax for better future compatibility
- Rewrite test_permissions_command.py to use integration tests instead of fragile mocks

HsiuChuanHsu committed:
feat(fab): add missing test_dag_permissions.py
Add unit tests for dag_permissions module to satisfy project structure
requirements

pierrejeambrun commented on providers/fab/src/airflow/providers/fab/auth_manager/cli_commands/permissions_command.py:
```diff
@@ -0,0 +1,122 @@
+                if resource.name.startswith(RESOURCE_DAG_PREFIX):
+                    dag_id = resource.name[len(RESOURCE_DAG_PREFIX) :]
+                elif resource.name.startswith("DAG Run:"):
+                    dag_id = resource.name[len("DAG Run:") :]
```
`RESOURCE_DAG_RUN_PREFIX` to avoid hardcoding `DAG Run:`

pierrejeambrun commented on providers/fab/src/airflow/providers/fab/auth_manager/cli_commands/permissions_command.py:
```diff
@@ -0,0 +1,122 @@
+                elif resource.name.startswith("DAG Run:"):
+                    dag_id = resource.name[len("DAG Run:") :]
+                elif resource.name.startswith("Task Instance:"):
+                    dag_id = resource.name[len("Task Instance:") :]
```
That doesn't exist I think.

pierrejeambrun commented on providers/fab/src/airflow/providers/fab/auth_manager/cli_commands/permissions_command.py:
```diff
@@ -0,0 +1,122 @@
+                    else f"clean up permissions for DAG '{args.dag_id}'"
+                )
+                confirm = input(f"\nDo you want to {action}? [y/N]: ")
+                if confirm.lower() not in ("y", "yes"):
```
We have a `TRUE_LIKE_VALUES` and `to_boolean` that could be helpful instead of doing that manually.

vincbeck commented on providers/fab/src/airflow/providers/fab/auth_manager/dag_permissions.py:
```diff
@@ -0,0 +1,110 @@
+#
```
I would move this code either in `providers/fab/src/airflow/providers/fab/auth_manager/cli_commands/permissions_command.py` or in a new file in `providers/fab/src/airflow/providers/fab/auth_manager/cli_commands/`

vincbeck commented on providers/fab/src/airflow/providers/fab/auth_manager/dag_permissions.py:
```diff
@@ -0,0 +1,110 @@
+        with create_session() as session:
+            _cleanup_dag_permissions_impl(dag_id, session)
+    else:
+        _cleanup_dag_permissions_impl(dag_id, session)
```
The decorator `@provide_session` does pretty much what you are doing here

vincbeck reviewed (commented):
Few comments but overall I really like it better that way! Thanks for the changes!

pierrejeambrun commented on providers/fab/tests/unit/fab/auth_manager/cli_commands/test_permissions_command.py:
These tests are not testing much beside the 'output print value'. We should also assert that `cleanup_dag_permissions` is being called with the appropriate params

pierrejeambrun commented on providers/fab/tests/unit/fab/auth_manager/test_dag_permissions.py:
Maybe we should mock less, and actually populate the db with some permissions, and verify what's being deleted.

pierrejeambrun reviewed (commented):
Looks good overall, a few suggestions.

HsiuChuanHsu committed:
refactor(fab): Consolidate DAG permissions code and modernize patterns
- Replace hardcoded resource strings with constants in permissions_command.py
  - Use RESOURCE_DAG_PREFIX and RESOURCE_DETAILS_MAP for DAG Run
  - Remove deprecated Task Instance resource handling
- Change to session handling with @provide_session decorator
- Replace manual boolean parsing with airflow.utils.strings.to_boolean
- Consolidate dag_permissions.py functionality into permissions_command.py
- Consolidate test files for better maintainability

HsiuChuanHsu committed:
test: enhance FAB permissions tests with function verification and real database operations
1. Function call verification (TestPermissionsCommand)
  - Replace stdout-only testing with precise function call verification
  - Add mock assertions for cleanup_dag_permissions with exact parameters
2. Real database operations (TestDagPermissions):
  - Replace mock-heavy approach with actual database interactions
  - Use real Resource, Action, and Permission entities with constraint handling

vincbeck reviewed (approved):
Nice!

pierrejeambrun reviewed (approved):
Nice thanks
